### ISSUE TITLE
Relation to NaN non-determinism

### ISSUE THREAD

**Issue by RalfJung** [WebAssembly/relaxed-simd#154]
This proposal emphasizes heavily that it introduces non-deterministic instructions. This makes it sound like non-determinism is a new concept that was otherwise not found in wasm. This is surprising to me since in my understanding, wasm already contains non-deterministic instructions. Specifically, the behavior of NaNs is [explicitly documented](https://webassembly.github.io/spec/core/exec/numerics.html#nan-propagation) as non-deterministic:

> Otherwise the payload is picked non-deterministically among all [arithmetic NaNs](https://webassembly.github.io/spec/core/syntax/values.html#arithmetic-nan); that is, its most significant bit is and all others are unspecified.

Am I misunderstanding something? If not, it might be worth clarifying that, and also clarifying whether these are "the same kind" of non-determinism or not.

**Comment by rossberg** [comment]
The text is perhaps somewhat confusing. What it's getting at is that this proposal introduces a new class of non-determinism that we did not have before, namely one that is more accurately understood as _implementation-dependent_ behaviour. With NaNs before, the choice is always independently arbitrary and can differ on every execution; the spec didn't guarantee any semantics that could vary across implementation. With relaxed SIMD there are some guarantees that the behaviour is basically deterministic for a given "environment" (which is a rather hand-wavy notion, as you note in #153).

In other words, the spec now officially supports and _blesses_ probing for and depending on "non-deterministic" behaviour. Of course, that is rather unfortunate, because it adds complexity to the specification and undermines portability, which is why it is called out.

**Comment by RalfJung** [comment]
> In other words, the spec now officially supports and blesses probing for and depending on "non-deterministic" behaviour. Of course, that is rather unfortunate, because it adds complexity to the specification and undermines portability, which is why it is called out.

Okay, I see. Thanks for the clarification. The first example in this proposal very much reads like the kind of non-determinism NaNs have, so I think it'd be good to clarify this. I guess it is too late to eschew "non-determinism" for some other term like "environment-dependent behavior"?

**Comment by rossberg** [comment]
Not too late to improve the explaner, but technically it still amounts to non-determinism, only of a more complicated form (vectors of sets of behaviours instead of just sets of behaviours, and the projection function is arbitrary).

**Comment by RalfJung** [comment]
> vectors of sets of behaviours

I admit I didn't get that part of the overview at all.

Using a `choice: () -> bool` function as an example since it is easier to discuss:

To me what you describe sounds like: "When `choice` executes, check a write-once global cache (of type `Option<bool>`) for whether we already made a choice. If yes, use that choice. If no, make a non-deterministic choice, record the result in the cache, and use that for this operation."

Or, equivalently: initialization of the Abstract Machine is non-deterministic. When the AM is initialized, non-deterministically pick a `bool` and store it in the global state. This part of the state is immutable from now on. Whenever `choice` is executed, return the value stored in the global state.

So yes, this "amounts to non-determinism", in the sense that non-determinism is somehow involved. But summarizing that as "`choice` is non-deterministic" is massively misleading! It's not wrong, but it's leaving away a ton of important aspects of `choice`. This operation described above is obviously very different from the normal non-deterministic choice, and using the same name for both is bound to cause confusion. (In Rust we'll probably call this "implementation-specific behavior" or something like that.)

But what does any of this have to do with vectors of sets of behaviors...?

**Comment by rossberg** [comment]
Yes, except that Wasm deliberately has no notion of "initialization of the abstract machine", and it's not at all clear what that would mean, i.e., when it would happen, or whether it could happen repeatedly, e.g., when code is persisted.

> But what does any of this have to do with vectors of sets of behaviors...?

The choice function returns an index into a vector of allowed non-deterministic behaviours. And in fact, we have separate choice functions for different groups of operations, with varying index range. It's all pretty messy...

**Comment by RalfJung** [comment]
> The choice function returns an index into a vector of allowed non-deterministic behaviours. 

You mean this: `IMPLEMENTATION_DEFINED_ONE_OF(0, INT32_MIN)`? That's just listing the options that we are "non-deterministically" picking from, it doesn't clarify in any way when and how that choice happens.

IOW, the "vector" seems to be just the usual list that comes out of the non-determinism monad.

Btw, the "modified spec" link at https://webassembly.github.io/relaxed-simd/ (found on the root README of this repo) is broken.

> Yes, except that Wasm deliberately has no notion of "initialization of the abstract machine", and it's not at all clear what that would mean, i.e., when it would happen, or whether it could happen repeatedly, e.g., when code is persisted.

Well that relates to the question about "the environment" then. That very much sounds like such an initialization in disguise -- the AM is initialized when "the environment" is fixed.

**Comment by rossberg** [comment]
No, that's not what it is. It rather works along the lines of
```
IMPLEMENTATION_DEFINED_ONE_OF([
  NONDETERMINISTIC_ONE_OF(X1, Y1),
  NONDETERMINISTIC_ONE_OF(X2, Y2),
  NONDETERMINISTIC_ONE_OF(X3, Y3),
])
Thus a vector of sets of possibilities. I truly wish it was simpler, and that we could have gotten away with just plain sets, but SIMD champions argued this extra guarantee was important.

I agree that the question of "fixing" the environment is rather fuzzy. In the respective discussions the conclusion was that code mobility is not an issue Wasm needs to be concerned with yet (which I suspect may be naive). As far as the observable semantics is concerned, the "environment" hence is fixed upfront and forever for any given program -- you just can't tell in advance what it is.

**Comment by RalfJung** [comment]
> No, that's not what it is. It rather works along the lines of

Oh I see, there are nested choices.

Which of these are "truly non-deterministic at runtime" and which are the ones that depend on "the environment"?

> the "environment" hence is fixed upfront and forever for any given program -- you just can't tell in advance what it is.

In other words, it is fixed when the AM is initialized? ;)

**Comment by rossberg** [comment]
> Where can I find `NONDETERMINISTIC_ONE_OF`?

That just is the form of nondeterminism we already have in Wasm 1.0/2.0.

Neither operator exists in the actual spec, I believe that's just an explanatory device in the overview.

(I just had another look at the relaxed SIMD spec draft, and to be honest, what it specifies is still lacking. It defines the vectors of non-deterministic sets for numeric primitives, but fails to connect that to the execution of instructions in a formal manner.)

> Is `NONDETERMINISTIC_ONE_OF` "truly non-deterministic at runtime" and `IMPLEMENTATION_DEFINED_ONE_OF` "depends on the environment"?

In reality, both may be implementation-dependent choices, but for the former there are zero guarantees of reproducability, while the latter choice is guaranteed to be fixed for each run (which is what causes all the headache).

> In other words, it is fixed when the AM is initialized? ;)
> 
> Since obviously I can run the same program in different environments and then it will see different choices.

Sure, but initialisation is not an observable step, nor is the AM itself an observable entity. You can pretend the choice was made with the big bang. Semantically, it's more appropriate to think of it as (a set of) global constants/parameters (one for each relaxed operator) whose values are unknown.

AFAICS, there is no meaningful way to make it more precise without hardcoding assumptions about Wasm that may not work for all implementations (e.g., any specific notion of engine initialisation).

@ngzhian, @Maratyszcza, can one of you fix the spec build on this repo, such that it shows up on github.io?

**Comment by rossberg** [comment]
> Can you point me at a place in the spec 

Well, as I just mentioned, a proper definition of this appears to be still missing from the draft spec. All I see so far are a couple of informal sentences [here](https://github.com/WebAssembly/relaxed-simd/blob/22257c57ba60f59d2b81a267635c5eba28719c0e/document/core/exec/numerics.rst#relaxed-operations), but we'll need more than that.
